File: java/examples/src/main/java/org/pactman/nonprofitcheckplus/examples/cases/Ex04NameComparison.java
package org.pactman.nonprofitcheckplus.examples.cases;

import org.pactman.nonprofitcheckplus.Sources;
import org.pactman.nonprofitcheckplus.devtools.FixtureEins;
import org.pactman.nonprofitcheckplus.examples.support.Example;
import org.pactman.nonprofitcheckplus.examples.support.ExampleContext;
import org.pactman.nonprofitcheckplus.examples.support.Matching;
import org.pactman.nonprofitcheckplus.examples.support.Output;
import org.pactman.nonprofitcheckplus.models.Nonprofit;
import org.pactman.nonprofitcheckplus.models.Pub78Source;

/**
 * EX-04 — Comparing an applicant's name against the record.
 *
 * <p>The API returns three names for one organization: the BMF name, the
 * Publication 78 name, and an "also known as". They disagree routinely, and
 * none of them is wrong. Comparing an applicant's typed name against them is a
 * customer policy question, which is why the comparison lives here and not in
 * the SDK.
 */
public final class Ex04NameComparison implements Example {

    @Override
    public String id() {
        return "ex-04";
    }

    @Override
    public String title() {
        return "Comparing an applicant's name against the IRS record";
    }

    @Override
    public void run(String[] args) {
        String applicant = args.length > 0 ? args[0] : "Meals Today Example Nonprofit, Inc.";

        try (ExampleContext context = ExampleContext.withFixtures()) {
            Nonprofit nonprofit = context.client()
                    .nonprofits()
                    .check(FixtureEins.PUBLIC_CHARITY)
                    .getNonprofit();

            Pub78Source pub78 = Sources.pub78(nonprofit);

            Output.heading("The names the API returned");
            Output.field("applicant typed", applicant);
            Output.field("organization_name", nonprofit.getOrganizationName());
            Output.field("bmf_organization_name", nonprofit.getBmfOrganizationName());
            Output.field(
                    "pub78_organization_name",
                    pub78 == null ? Output.ABSENT : pub78.getOrganizationName());
            Output.field("organization_name_aka", nonprofit.getOrganizationNameAka());

            Output.heading("Normalized for comparison");
            Output.field("applicant", Matching.normalizeName(applicant));
            Output.field("organization_name", Matching.normalizeName(nonprofit.getOrganizationName()));

            Output.heading("How each one compares");

            for (String[] candidate : new String[][] {
                {"organization_name", nonprofit.getOrganizationName()},
                {"bmf_organization_name", nonprofit.getBmfOrganizationName()},
                {"pub78_organization_name", pub78 == null ? null : pub78.getOrganizationName()},
                {"organization_name_aka", nonprofit.getOrganizationNameAka()},
            }) {
                Matching.Comparison comparison =
                        Matching.compareNames(applicant, candidate[1]);
                Double overlap = Matching.wordOverlap(applicant, candidate[1]);

                Output.field(
                        candidate[0],
                        comparison + (overlap == null
                                ? ""
                                : String.format(" (word overlap %.2f)", overlap)));
            }
        }

        Output.note("A missing name is UNCOMPARABLE, never agreement — scoring an absent value\n"
                + "as a match is how a blank record passes a name check. And a name that\n"
                + "differs is not fraud: organizations rename, and IRS records lag. Route the\n"
                + "disagreement to a reviewer; do not fail the applicant on it.");
    }
}

File: ruby/examples/ex_04_name_comparison.rb
# frozen_string_literal: true

# EX-04 — Applicant name comparison.
#
# An applicant types a name during onboarding. The API returns the name IRS
# records hold, plus an alternate name when one exists. Punctuation, casing and
# abbreviation differences are normal; they are not evidence of fraud.
#
# The SDK deliberately has no `names_match?`. What counts as a match is your
# policy, so the comparison lives here, in customer code.
#
# Run:  PACTMAN_API_KEY=... bundle exec ruby examples/ex_04_name_comparison.rb

require_relative "lib/client"
require_relative "lib/fixture_api"
require_relative "lib/matching"
require_relative "lib/print"

include Print

client = ExampleClient.create_client
public_charity = FixtureApi::EINS[:public_charity]

# Three applicants against the same organization: a formatting difference, an
# abbreviation difference, and a genuinely different name.
applicants = [
  { ein: public_charity, legal_name: "Meals Today Example Nonprofit" },
  { ein: public_charity, legal_name: "meals today example nonprofit, inc." },
  { ein: public_charity, legal_name: "Springfield Animal Rescue" }
]

# Your routing policy, not the SDK's.
def route_name_outcome(outcome)
  case outcome
  when "exact", "normalized"
    "continue — the submitted name agrees with an IRS-held name"
  when "not_returned"
    "manual review — the API returned no name to compare against"
  else
    "manual review — a human decides whether this is a rebrand, a typo, or the wrong EIN"
  end
end

applicants.each do |applicant|
  nonprofit = client.nonprofits.check(applicant[:ein]).nonprofit

  if nonprofit.nil?
    puts "No record for #{applicant[:ein]}."
    next
  end

  comparison = Matching.compare_name(
    applicant[:legal_name],
    "organization_name" => nonprofit.organization_name,
    "organization_name_aka" => nonprofit.organization_name_aka
  )

  heading "Applicant: #{applicant[:legal_name]}"
  field "organization_name", nonprofit.organization_name
  field "organization_name_aka", nonprofit.organization_name_aka
  field "normalized applicant", comparison.submitted

  comparison.candidates.each do |candidate|
    bullet "#{candidate.source} normalizes to \"#{candidate.normalized}\""
  end

  field "outcome", comparison.outcome
  field "matched field", comparison.matched_field
  field "agreement", Matching.agreement?(comparison.outcome)
  field "routed to", route_name_outcome(comparison.outcome)
end

note "A name mismatch is a reason to look, not a finding. Organizations rebrand, file\n" \
     "under a parent, and appear in IRS data under a name no donor would recognize.\n" \
     "This example routes disagreement to review; it never labels an applicant."
